handlers/handlers.go

```diff
@@ -48,3 +49,61 @@ func DoCheckCookie(h http.Handler, key Key) http.Handler {
 		h.ServeHTTP(w, req)
 	})
 }
+
+/* My head hurts...
+ControllerHandler is called from an app which then creates a function of type ControllerHandlerFunc which returns a http.Handler (and error)
```

Comment:
I think you are on the right track. I believe we would want the `ControllerHandlerMiddleware` called from the app passing in a `ControllerHandlerFunc` as an argument, then the `ControllerHandlerMiddleware` should return a closure that calls the `ControllerHandlerFunc` that was passed in.

Something like:

```
func ControllerHandlerMiddleware(h ControllerHandlerFunc) (http.Handler) {
    return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
        locale := getLocaleCode(req)
        collectionID := getCollectionIDFromContext(req)
        accessToken := getUserAccessTokenFromContext(req)

        h(w, r, locale, collectionID, accessToken)
    })
}
```